### DISCUSSION TITLE
gg module: odd create_image functions situation

### DISCUSSION THREAD

**Discussion by Wajinn** [vlang/v#16754] [General]
First, happy holidays and hope all are doing well.  Happen to be playing with some examples, and came across the [flappylearning](https://github.com/vlang/v/tree/master/examples/flappylearning) one.  This example works fine on Windows, but simply would not work on Android (after using vab). When looking a bit deeper as to what could be wrong, came across a number of posts detailing the issue.  It appears that the gg module [create_image](https://modules.vlang.io/gg.html#Context.create_image) functions issue has been an ongoing thing. It looks like it may have worked previously, possibly sometimes, and not other times.  Some of the links that are connected; [16542](https://github.com/vlang/v/discussions/16542), [239](https://github.com/vlang/vab/issues/239), [16564](https://github.com/vlang/v/pull/16564)...

The thing about the issue is that since it is an example, probably many would expect it to work and would be disappointed if it didn't.  Something else that I noticed is the way in which flappylearning was attempting to load the image seems a bit strange.  Would think it could load it from a file, like how fonts are, but instead it was using from_byte_array. Just saying and noticing the difference.  By the way, neither option seems to presently work.

Wondering if there are those that got it to work or have any more in depth knowledge about the situation?

**Comment by larpon** [comment]
Happy holidays.
It works on my Android device after #16564 , and to my knowledge no one have had any unexpected problems since? The image viewer example works also. Is this a bug report? For details please see the wall of text reasoning I give in #16564 (more investigation in #16670). It's a small friction between the the channel count in sokol_gfx and the stbi image loader - something that existed because gg is trying to cover more ground than either of the two libs previously mentioned.

(gg tries to hide away any pains from the user regarding image channels aka. Red, Green, Blue and Alpha channels. Images doesn't always have all channels set - especially if an image has been optimised by some external app etc., and users shouldn't be stopped by such irrelevant problems that can arise,  when simply trying to load and display an image)

**Reply by larpon** [reply]
So if the example doesn't work on your Android device, please make sure V and vab is updated, and make sure you're not building from cached content i.e. pass `--nocache` to vab

**Reply by larpon** [reply]
also make sure you're not running a previously installed app that was broken. So uninstall any previous attempts at running flappylearning on your device etc.

**Reply by Wajinn** [reply]
Thanks for the replies.

> to my knowledge no one have had any unexpected problems since? 

Well, I'm someone :-)

> please make sure V and vab is updated

Yes, they are.  Using latest V (`V 0.3.2 6a179a2`) and vab (vab `0.3.7 a835220`). Have compiled and run other examples successfully (2048 and tetris). Issue appears specific to flappylearning, and more specifically the `create_image` functions . Other experimental programs will install successfully with `create_image`, but not show an image.  With flappylearning, `create_image_from_byte_array` and `read_apk_asset`, will install successfully on android, but program will crash on attempting to play.  There is no error screen, just go black, and will not run. 

However, I'm using an older version of android (Oreo) with `api 26`, as like to play with that version and then work up.  Of course will do some more experimenting. 

What version of android and api to you happen to be running?

Maybe others can chime in on if they got flappylearning running or not.  Kind of curious, thus the discussion.

**Reply by larpon** [reply]
I'm always targeting API 21 as a minimum SDK  version via `--min-sdk-version` and then the highest API to build against in my env which is currently 31 (`--api 31`, usually auto-detected).

Can you open a bug report in vlang/v and make sure you include all details such as `vab doctor` output and *the exact vab command* to reproduce the broken build. I'll see if I can remember to send you an APK of flappy bird built on my setup sometime after the holidays, and then see if that can run on your device - to rule out any Android vendor specific etc.

**Reply by larpon** [reply]
I hope we can reproduce it locally, so we can put an end to all these indifferences

**Reply by larpon** [reply]
Also include the full source code of what you're trying to build in bug report, it's important in this case

**Reply by larpon** [reply]
... a bug report in vlang/vab is also fine 🙂

**Reply by larpon** [reply]
Also, this might be a relevant thing to try:
https://github.com/vlang/v/issues/16670#issuecomment-1350225967

**Reply by Wajinn** [reply]
> Also include the full source code of what you're trying to build in bug report, it's important in this case

I was simply taking the existing examples, stripping out extraneous code, and making them simpler to better isolate issues. Example below:
```
module main

import gg
import gx
import os

const (
	win_width  = 500
	win_height = 512
)

struct App {
	mut:
	gg               &gg.Context = unsafe { nil }
	background       gg.Image
	width            f64 = win_width
	height           f64 = win_height
}

fn main() {
	mut app := &App {
		gg: 0
	}
	app.gg = gg.new_context(
		bg_color: gx.white
		width: win_width
		height: win_height
		create_window: true
		window_title: 'flappylearning-v'
		frame_fn: frame
		user_data: app
		init_fn: init_images
	)
	app.gg.run()
}


fn init_images(mut app App) {
	$if android {
		background := os.read_apk_asset('img/background.png') or { panic(err) }
		app.background = app.gg.create_image_from_byte_array(background)
	} 
	$else {
		app.background = app.gg.create_image(os.resource_abs_path('assets/img/background.png'))
	}
}

fn frame(app &App) {
	app.gg.begin()
	app.draw()
	app.gg.end()
}

fn (app &App) draw() {
	app.gg.draw_image(0, 0, app.gg.width, app.gg.height, app.background)
}
```

**Reply by Wajinn** [reply]
> ... a bug report in vlang/vab is also fine slightly_smiling_face

Vab doesn't appear to be the issue, or at least I'm having trouble isolating what is.  I would rather create an simplified example that specifically isolates the problem, as to be clear what's going on. For instance, I was able to get all examples (including the simplified one posted) to compile with vab without any problems.  And the simple ones run correctly on android, as they should.  

The problems appear to have something to do with, `os.read_apk_asset`, in more complicated code and related to also `init_fn` .  In simpler code, as posted, things work fine.  But, for unknown reasons in more elaborate code, graphic files don't seem to load and such programs crash the android app.  In the android logcat, errors like `gg: draw_image() bad img id 0 (img cache len = 0)` and `libEGL called unimplemented OpenGL ES API` could be seen.  But have to go back into details, as which test matched up to what.  Also, without loading graphic files, things can work fine.  

Additionally, it appears that only `create_image` can not be used for android (as oppose to `create_image_from_byte_array`).  I suppose I'm also wondering if others, using vlang, have alternate ways of loading graphic files for android, besides only the `os.read_apk_asset` and  `create_image_from_byte_array ` combination (so like the discussion format).  Alternative suggestions welcomed.

As the holidays and upcoming new year, sure people are partying, drinking, movie binge watching, etc...  So look to more fun playing around with this soon afterwards.

**Reply by larpon** [reply]
Highly strange 😅 thanks for getting back

**Reply by larpon** [reply]
@Wajinn - `os.read_apk_asset` is quite dependent on the resource actually being *in the APK/AAB* - e.g. it might be a an APK package issue rather than any `gg` related. For V's examples specifically `vab` includes resources automatically. Can you please try and add the path to the resource/image/file you are trying to load in your app with `--assets`. Example: `vab --assets /path/to/assets ...` - and afterwards make sure that your APK actually contains the assets (APK's are just modified .zip files so your should be able to open and inspect it with most "normal" ZIP tools)

**Reply by larpon** [reply]
... as for `create_image` vs. `create_image_from_byte_array` - Android's filesystem doesn't work like you would expect it to do on a normal desktop PC, because of Android's resource system, security policies and the fact that APK/AAB allows users to package their own files - everything file related for APK/AABs needs to be handled specially. See <https://developer.android.com/reference/android/content/res/AssetManager> for the functionality we also use in V/`vab`

**Reply by spytheman** [reply]
imho a small `dump( background#[0..10] )` call will help to show if the file was loaded from the APK

**Reply by larpon** [reply]
@Wajinn here's a version of flappylearning built with: `vab --min-sdk-version 21 --api 26 -o /tmp/flappylearning.apk -v 3 ~/path/to/v/examples/flappylearning/`

You should be able to unzip it and install the `flappylearning.apk` file in the archive, on your device with either `vab` or via file transfer.
[flappylearning.zip](https://github.com/vlang/v/files/10329086/flappylearning.zip)

If this does not work on your device we'd really like to hear back from you about what the *exact* log output is.

**Reply by Wajinn** [reply]
> `os.read_apk_asset` is quite dependent on the resource actually being _in the APK/AAB_ - e.g. it might be a an APK package issue rather than any `gg` related. 

> imho a small dump( background#[0..10] ) call will help to show if the file was loaded from the APK

Of course you guys are sharp foxes with great advice.  Sorry for being a bit late getting back, holiday hangover.  As mentioned, the simplified examples worked, indicating no trouble with `gg`.  The issues were a combination of incorrect usage, in more elaborate variations based on the examples, with `os.abs_path` and `os.join_path` inside the `os.read_apk_asset`(inside here) function, not using `os.read_apk_asset` for android, or the correct file not being in the asset. So, its all sorted out. Thanks.

**Reply by larpon** [reply]
> So, its all sorted out. Thanks.

That's good to hear, thanks for letting us know.
